File: interview_cake/arrays_and_strings/reverse_string.js
function reverse(arrayOfChars) {

  // Reverse the input array of characters in place
  let leftIndex = 0;
  let rightIndex = arrayOfChars.length - 1;
  
  while (rightIndex > leftIndex) {
    placeholder = arrayOfChars[leftIndex];
    arrayOfChars[leftIndex] = arrayOfChars[rightIndex];
    arrayOfChars[rightIndex] = placeholder;
    leftIndex += 1;
    rightIndex -= 1;
  }
  
  return arrayOfChars;

}


















// Tests

let desc = 'empty string';
let input = ''.split('');
reverse(input);
let actual = input.join('');
let expected = '';
assertEqual(actual, expected, desc);

desc = 'single character string';
input = 'A'.split('');
reverse(input);
actual = input.join('');
expected = 'A';
assertEqual(actual, expected, desc);

desc = 'longer string';
input = 'ABCDE'.split('');
reverse(input);
actual = input.join('');
expected = 'EDCBA';
assertEqual(actual, expected, desc);

function assertEqual(a, b, desc) {
  if (a === b) {
    console.log(`${desc} ... PASS`);
  } else {
    console.log(`${desc} ... FAIL: ${a} != ${b}`);
  }
}

File: interview_cake/arrays_and_strings/reverse_string.rb
def reverse!(string)
  left_index = 0
  right_index = string.length - 1
  
  while right_index > left_index
    placeholder = string[left_index]
    string[left_index] = string[right_index]
    string[right_index] = placeholder
    left_index += 1
    right_index -= 1
  end
  
  string

end


















# Tests

def run_tests
  desc = 'empty string'
  string = ''
  reverse!(string)
  expected = ''
  assert_equal(string, expected, desc)

  desc = 'single character string'
  string = 'A'
  reverse!(string)
  expected = 'A'
  assert_equal(string, expected, desc)

  desc = 'longer string'
  string = 'ABCDE'
  reverse!(string)
  expected = 'EDCBA'
  assert_equal(string, expected, desc)
end

def assert_equal(a, b, desc)
  puts "#{desc} ... #{a == b ? 'PASS' : "FAIL: #{a.inspect} != #{b.inspect}"}"
end

run_tests
